python/ffsim/tenpy/circuits/gates.py

```diff
@@ -0,0 +1,451 @@
+        NNgate = NNgate_a @ NNgate_b
+    else:
+        raise ValueError("undefined spin")
+
+    # convert to (N, Sz)-symmetry-conserved basis
+    NNgate_sym = sym_cons_basis(NNgate)
+
+    return NNgate_sym
+
+
+def apply_gate1(psi: MPS, U1: np.ndarray, site: int) -> None:
+    r"""Apply a single-site gate to a
+    `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+    wavefunction.
+
+    Args:
+        psi: The `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+        U1: The single-site quantum gate.
+        site: The gate will be applied to `site` on the
+            `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+
+    Returns:
+        None
+    """
+
+    # apply single-site gate
+    U1_npc = npc.Array.from_ndarray(U1, [shfs.leg, shfs.leg.conj()], labels=["p", "p*"])
+    psi.apply_local_op(site, U1_npc)
+
+
+def apply_gate2(
+    psi: MPS,
+    U2: np.ndarray,
+    sites: tuple[int, int],
+    *,
+    eng: TEBDEngine,
+    chi_list: list,
+    norm_tol: float = 1e-5,
+) -> None:
+    r"""Apply a two-site gate to a `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+    wavefunction.
+
+    Args:
+        psi: The `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+        U2: The two-site quantum gate.
+        sites: The gate will be applied to adjacent sites `(site1, site2)` on the
+            `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+        eng: The
+            `TeNPy TEBDEngine <https://tenpy.readthedocs.io/en/latest/reference/tenpy.algorithms.tebd.TEBDEngine.html#tenpy.algorithms.tebd.TEBDEngine>`__.
+        chi_list: The list to which to append the MPS bond dimensions as the circuit is
+            evaluated.
+        norm_tol: The norm error above which we recanonicalize the wavefunction, as
+            defined in the
+            `TeNPy documentation <https://tenpy.readthedocs.io/en/latest/reference/tenpy.algorithms.dmrg.DMRGEngine.html#cfg-option-DMRGEngine.norm_tol>`__.
+
+    Returns:
+        None
+    """
+
+    # check that sites are adjacent
+    if abs(sites[0] - sites[1]) != 1:
+        raise ValueError("sites must be adjacent")
+
+    # check whether to transpose gate
+    if sites[0] > sites[1]:
+        U2 = U2.T
+
+    # apply NN gate between (site1, site2)
+    U2_npc = npc.Array.from_ndarray(
+        U2, [shfsc, shfsc.conj()], labels=["(p0.p1)", "(p0*.p1*)"]
+    )
+    U2_npc_split = U2_npc.split_legs()
+    eng.update_bond(max(sites), U2_npc_split)
+    chi_list.append(psi.chi)
```

Comment:
This line is the first line in the function that seems to use `psi`. Is `psi` used implicitly before this? Is it stored in the engine? Sorry, I am not familiar with TenPY.

Reply:
Yes, the line `eng.update_bond` implicitly updates `psi`

Reply:
So `eng` contains a reference to `psi` then? Does `psi` also contain a reference to `eng`? If either of these is the case, then we only need to pass one of them to the function, right?

Reply:
The `TEBDEngine` is designed to run the TEBD algorithm (this was originally imagined to be some lengthy time evolution, not simply the application of a single gate). As inputs, it takes the wavefunction (`psi`), the model, and some options. After the engine has run, `psi` will have been updated in-place. Although the engine knows about `psi`, there is no public-facing attribute to get `psi`. The closest we have is [get_resume_data](https://tenpy.readthedocs.io/en/latest/reference/tenpy.algorithms.tebd.TEBDEngine.html#tenpy.algorithms.tebd.TEBDEngine.get_resume_data), which was originally intended to be a checkpoint when running some lengthy time evolution on a cluster. This function uses the internal attribute `self._resume_psi` from the `Algorithm` class to get `psi`, I think. If you're happy using internal attributes, then `eng` contains a reference to `psi`. However, `psi` does not know about `eng`.

Reply:
TeNPy was not designed with quantum circuits in mind :sweat_smile: